### Graded Challenge 3

### Nama: Stanly Sukmajaya

### Batch: RMT-036

### Program ini dibuat untuk menganalisa produk-produk seblak yang dijual di Tokopedia. Data-data akan di scraping secara mandiri, dan akan diambil datanya minimal 50 data. Akhirnya, data akan dianalisa dan diberi kesimpulan.


In [1]:
# Import selenium
from selenium import webdriver

In [27]:
from bs4 import BeautifulSoup
import time
driver = webdriver.Chrome()

names = []
prices = []
items_sold = []
ratings = []
sellers = []
cities = []
# Looping URL untuk mendapatkan data di 10 pages
for i in range(1, 11):
  url = f"https://www.tokopedia.com/search?navsource=&page={i}&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st="
  driver.get(url)
  html = driver.page_source

  page = BeautifulSoup(html, "html.parser")
  time.sleep(10)
  result = page.find("div", {"class": "css-jza1fo"})
  items = result.find_all("div", {"class": "KeRtO-dE1UuZjZ6nmihLZw=="})
  # Looping data div dengan class 'KeRtO-dE1UuZjZ6nmihLZw=='
  for item in items:
    name = item.find("span", {"class": "OWkG6oHwAppMn1hIBsC3pQ=="})
    if name and name.get_text():
      names.append(name.get_text())
    else:
      names.append(None)
    price = item.find("div", {"class": "_8cR53N0JqdRc+mQCckhS0g=="})
    if price and price.get_text():
      prices.append(price.get_text())
    else:
      another_price = item.find("span", {"class": "en+9Xhk5rmGNLiUfSuIuqg=="})
      if another_price and another_price.get_text():
        prices.append(another_price.get_text())
      else:
        prices.append(None)
    sold = item.find("span", {"class": "eLOomHl6J3IWAcdRU8M08A=="})
    if sold and sold.get_text():
      items_sold.append(sold.get_text())
    else:
      items_sold.append(None)
    rating = item.find("span", {"class": "nBBbPk9MrELbIUbobepKbQ=="})
    if rating and rating.get_text():
      ratings.append(rating.get_text())
    else:
      ratings.append(None)
    seller_and_city = item.find_all("span", {"class": "flip"})
    for index, data in enumerate(seller_and_city):
      if index == 0:
        if data and data.get_text():
          sellers.append(data.get_text())
        else:
          sellers.append(None)
      elif index == 1:
        if data and data.get_text():
          cities.append(data.get_text())
        else:
          cities.append(None)

print(names)
print(prices)
print(items_sold)
print(ratings)
print(sellers)
print(cities)

['KRUPUK SEBLAK RAFAEL | SEBRING | KERUPUK MAWAR PEDAS DAUN JERUK 250 GR', 'Seblak Rafael/Seblak Coet Instan Halal', 'Seblak Sultan instant', 'Kerupuk Seblak Rafael / Seblak Pedas 250 Gram - Tam Snack', 'KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS INDRAMAYU BISA BUAT SEBLAK', 'Kerupuk Mentah 250gr Seblak Mentah', 'Kylafood Seblak Rempah Authentik', 'seblak krupuk 5000 ,pedas dan gurih', 'Kerupuk Seblak Beton As Kering Cikruh 500 Gr Bumbu Organik Pedas Gurih', 'KERUPUK MENTAH SERBA 5000 BAHAN SEBLAK', 'Gelifood Combo 3pcs Seblak Ceker Tanpa Tulang Super Pedas', 'Kylafood Seblak Cup', 'pangsit segitiga kering atau cipet isi 100 toping seblak baso aci', 'Seblak Instan Premium Bangsatria', 'siomay mini kering 1kg toping seblak baso aci', 'KERUPUK JABLAY BANTET SEBLAK PELANGI', 'seblak mercon pedas galak', 'seblak instant sajodo snak', 'Kylafood Paket Seblak Cup isi 4 pcs', 'SEBLAK INSTAN MOMMY', 'Kylafood Paket Seblak Cup isi 4 pcs', 'Seblak Rafael Seblak Beton', 'Kylafood Seblak Mie Baso S

In [96]:
import pandas as pd

print(len(names))
print(len(prices))
print(len(items_sold))
print(len(ratings))
print(len(sellers))
print(len(cities))
# Masukkan data yang telah discraping ke dalam dataframe
tokopedia = pd.DataFrame()
tokopedia['name'] = names
tokopedia['price'] = prices
tokopedia['items_sold'] = items_sold
tokopedia['rating'] = ratings
tokopedia['seller'] = sellers
tokopedia['city'] = cities

# Simpan data yang telah discrape ke dalam csv
tokopedia.to_csv("original_data.csv")

100
100
100
100
100
100


In [133]:
# Check data-data yang null, dan isi dengan string "0"
print(tokopedia.isna().sum())
tokopedia = tokopedia.fillna("0")

# Rapikan data-data items_sold, price dan rating dan ubah tipe-tipe data tersebut menjadi float dan int.
tokopedia['items_sold'] = tokopedia['items_sold'].str.replace("rb", "000")
tokopedia['items_sold'] = tokopedia['items_sold'].str.replace("terjual", "")
tokopedia['items_sold'] = tokopedia['items_sold'].str.replace("+", "")
tokopedia['items_sold'] = tokopedia['items_sold'].str.strip()
tokopedia['price'] = tokopedia['price'].str.replace("Rp", "")
tokopedia['price'] = tokopedia['price'].str.replace(".", "")
tokopedia[['price', 'rating']] = tokopedia[['price', 'rating']].astype(float)
tokopedia['items_sold'] = tokopedia['items_sold'].astype(int)

name          0
price         0
items_sold    3
rating        5
seller        0
city          0
dtype: int64


In [106]:
# Simpan data yang telah dirapikan ke dalam "cleaned_data.csv"
tokopedia.to_csv("cleaned_data.csv")
tokopedia

name    price  items_sold  \
0   KRUPUK SEBLAK RAFAEL | SEBRING | KERUPUK MAWAR...  21630.0          70   
1              Seblak Rafael/Seblak Coet Instan Halal  26600.0         500   
2                               Seblak Sultan instant  14000.0         250   
3   Kerupuk Seblak Rafael / Seblak Pedas 250 Gram ...  16000.0          70   
4   KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS IND...  21500.0         250   
..                                                ...      ...         ...   
95                  seblak kerupuk kering level pedas   5000.0          26   
96  kerupuk seblak kering mangar manggar jaat akar...  11250.0           6   
97    kerupuk seblak pedas daun jeruk kemasan 70 gram   4800.0          27   
98  NAAGIH! Seblak Cuanki Baso Aci Boci Instant Ku...   8000.0         100   
99  1KG Krupuk Seblak Kering Pedas/ Krupuk Seblak ...  21500.0          60   

    rating                  seller             city  
0      5.0     Toko kue Sumber Mas    Jakarta Timur  
1      4.9              Foodstocks  Jakarta Selatan  
2      4.9        Sultan Meat Shop            Medan  
3      5.0               TAM SNACK            Depok  
4      4.9                dariimyu   Kab. Indramayu  
..     ...                     ...              ...  
95     4.1  Permen Jahe Pekalongan    Kab. Pemalang  
96     4.8               Qie snack      Kab. Jember  
97     5.0       Gudang Snack Solo        Surakarta  
98     5.0              Ebay Store   Kab. Tangerang  
99     4.9             Vikasnack02    Kab. Karawang  

[100 rows x 6 columns]

In [107]:
# Lihat summary keseluruhan data
tokopedia.describe()

price   items_sold      rating
count    100.000000   100.000000  100.000000
mean   21268.310000   416.750000    4.613000
std    16945.683401   786.544037    1.078032
min     1100.000000     0.000000    0.000000
25%     9749.750000    30.000000    4.800000
50%    18449.500000   100.000000    4.900000
75%    26500.000000   500.000000    5.000000
max    91000.000000  4000.000000    5.000000

### SMART Framework

- Specific: Increase average items_sold to be 800 by launching more effective advertising campaigns and giving promotions & discounts to users.
- Measurable: Increase average items_sold from the current 416 to 800.
- Achievable: Based on the current data, it is showing great results, as the mean rating is 4.6, meaning users are satisfied, so achieving average items_sold to 800 is achievable.
- Relevant: Improving average items_sold is crucial to driving customer traffic, improving Tokopedia's reputation and increasing revenue. Moreover, seblak
- Time-bound: The goal is set for the next 1 year.

##### The goal is set for the next 1 year, by launching more effective advertising campaigns and giving promotions & discounts to users, the average items_sold of the product `seblak` is expected to reach 800.


In [109]:
# Lihat overview data-data yang int dan float untuk mendapatkan gambaran/high overview mengenai data-data tersebut.
new_tokopedia = tokopedia[['price', 'items_sold', 'rating']].copy()

summary = pd.DataFrame({
  'mean': new_tokopedia.mean(),
  'median': new_tokopedia.median(),
  'std': new_tokopedia.std(),
  'skewness': new_tokopedia.skew(),
  'kurtosis': new_tokopedia.kurtosis()
})

summary

mean   median           std  skewness   kurtosis
price       21268.310  18449.5  16945.683401  1.973404   4.878121
items_sold    416.750    100.0    786.544037  3.109596  10.145723
rating          4.613      4.9      1.078032 -4.018915  14.907508

## D. Analysis

### 1. Hitung rata-rata, median, standar deviasi, skewness, dan kurtosis dari kolom harga, banyak produk terjual, dan rating. Dari hasil perhitungan, insight apa saja yang bisa kamu dapatkan khususnya terkait dengan produk seblak dan datanya (distribusi dan kecenderungan ada/tidaknya outlier)? Note: Jika menemukan adanya indikasi outlier dari perhitungan ini, tidak perlu dihandle karena sifatnya alami. Dibiarkan saja.


### Dari data di atas, dapat dilihat bahwa semua nilai skew nya lebih dari 1, atau kurang dari -1. Ini menandakan bahwa di semua kolom terdapat nilai ekstrim. Pada kolom price dan items_sold, karena nilai skew nya lebih dari 1, menandakan bahwa outliersnya lebih condong ke kanan, dan kebanyakan data berada di sisi kiri distribusi, sedangkan pada kolom rating, karena nilai skew nya kurang dari -1, menandakan bahwa outliersnya lebih condong ke kiri, dan kebanyakan data berada di sisi kanan distribusi.


### Bisa disimpulkan dari analisa tersebut, bahwa untuk price dan items_sold lebih banyak barang yang terjual lebih murah dan lebih sedikit, sedangkan untuk rating, lebih banyak barang yang memiliki rating yang mendekati 5, dari pada yang mendekati 4.


### 2. Kamu memikirkan berapa potensi minimum dan maksimum pendapatan jika kamu menjual produk seblak?(Gunakan confidence interval untuk mendapatkan lower value dan upper value dari distribusi populasi pendapatan).

### Anggap data terdistribusi normal dan informasi produk terjual merupakan penjualan produk per bulan. Ambil confidence level 95%.


In [123]:
# Melakukan kalkulasi perkalian items_sold dan price untuk mendapatkan nilai revenue
new_tokopedia['revenue'] = new_tokopedia['items_sold'] * new_tokopedia['price']
new_tokopedia

price  items_sold  rating     revenue
0   21630.0          70     5.0   1514100.0
1   26600.0         500     4.9  13300000.0
2   14000.0         250     4.9   3500000.0
3   16000.0          70     5.0   1120000.0
4   21500.0         250     4.9   5375000.0
..      ...         ...     ...         ...
95   5000.0          26     4.1    130000.0
96  11250.0           6     4.8     67500.0
97   4800.0          27     5.0    129600.0
98   8000.0         100     5.0    800000.0
99  21500.0          60     4.9   1290000.0

[100 rows x 4 columns]

In [139]:
# Mendapatkan lower value dan upper value dari kolom revenue, dengan 95% confidence interval
from scipy import stats
import numpy as np
revenue_std = new_tokopedia['revenue'].std()
N = len(new_tokopedia['revenue'])
low, up = stats.norm.interval(0.95, loc=new_tokopedia['revenue'].mean(), scale=revenue_std/np.sqrt(N))

print(f"Lower value: {low:.2f}")
print(f"Upper value: {up:.2f}")

Lower value: 4134972.43
Upper value: 10160977.29


### 3. Apakah harga barang di Jabodetabek dan di luar Jabodetabek berbeda? mengingat biaya bahan baku di kedua lokasi berbeda. (Gunakan uji hipotesis yang diawali dengan menuliskan hipotesis null dan alternatifnya serta tentukan jenis hipotesis yang digunakan).


In [136]:
# Pisahkan data antara yang di daerah jabodetabek dan non-jabodetabek
tokopedia['city'].unique()
jabodetabek_list = ["Jakarta Timur", 
                    "Jakarta Selatan", 
                    "Jakarta Utara", 
                    "Jakarta Barat",
                    "Depok", 
                    "Kab. Bogor", 
                    "Tangerang", 
                    "Kab. Tangerang", 
                    "Tangerang Selatan"
                    ]

jabodetabek = tokopedia[tokopedia['city'].isin(jabodetabek_list)]
non_jabodetabek = tokopedia[~(tokopedia['city'].isin(jabodetabek_list))]

## h0: Harga seblak di Jabodetabek >= harga seblak di luar Jabodetabek

## h1: Harga seblak di Jabodetabek < harga seblak di luar Jabodetabek


In [138]:
# Lakukan Hypothesis testing terhadap harga-harga barang di jabodetabek dan non-jabodetabek
t_stat, p_val = stats.ttest_ind(jabodetabek['price'], non_jabodetabek['price'])
print('T-Statistic:', t_stat)
print('P-value:', p_val / 2) # Di bagi 2 karena one sided p value

T-Statistic: -0.4768320994448604
P-value: 0.31727162265625564


### Dari hasil yang didapat di atas, yaitu P-value (0.31) lebih dari critical value (0.5), sehingga h0 dapat diterima, dan kita menolak h1.Harga barang di daerah Jabodetabek sama atau lebih dari harga barang di luar Jabodetabek.


### 4. Apakah orang lebih cenderung suka dengan produk yang harganya murah?

### Kamu dapat jawab pertanyaan ini dengan uji korelasi. Gunakan library scipy jangan pandas. Analisis nilai korelasi dan p-value nya. Gunakan teknik yang sesuai dengan kondisi data!


In [135]:
# Lakukan correlation testing antara data-data yang harganya murah dan yang tidak murah
cheap_price = tokopedia['price'].quantile(0.5)
cheap_data = tokopedia[tokopedia['price'] <= cheap_price]
not_cheap_data = tokopedia[tokopedia['price'] > cheap_price]

correlation, pvalue = stats.spearmanr(cheap_data['items_sold'], not_cheap_data['items_sold'])
print("Correlation:", correlation)
print("P Value:", pvalue)
print("Cheap data items sold mean:", cheap_data['items_sold'].mean())
print("Not cheap data items sold mean:", not_cheap_data['items_sold'].mean())

Correlation: 0.0890880899920595
P Value: 0.5383962240782136
Cheap data items sold mean: 470.4
Not cheap data items sold mean: 363.1


### Berdasarkan data di atas, bisa disimpulkan bahwa tidak ada korelasi/relationship yang meaningful dari untuk items_sold dari data yang price nya rendah dan yang price nya tinggi. Namun, nilai rata-rata items_sold pada seblak yang harganya murah adalah 107 lebih dari seblak yang harganya mahal. Dari sini bisa disimpulkan bahwa orang lebih cenderung suka dengan produk yang harganya murah.


## E. Conclusion

### Tuliskan di markdown kesimpulan dari hasil analisis dari nomor 1-4. Kesimpulan yang baik menjawab tujuan yang ingin dicapai. Kamu dibebaskan menuliskan dalam format paragraf atau poin.


### Kesimpulan dari hasil-hasil nomor 1 sampai 4, menunjukkan bahwa sebagian customer puas dengan produk-produk seblak yang ditawarkan di Tokopedia, karena nilai rata-rata rating nya adalah 4.6. Untuk price, items_sold dan rating masing-masing memiliki outliers, dengan price dan items_sold outliersnya di arah kanan distribusi, sedangkan rating outliersnya di arah kiri distribusi. Untuk price dan items_sold, lebih banyak data yang value nya di kiri/rendah, sedangkan untuk rating lebih banyak data yang valuenya di kanan/tinggi. Potensi pendapatan jika saya mau jualan seblak, berdasarkan data adalah Rp 4,134,972.43 untuk nilai paling rendah, dan Rp 10,160,977.29 untuk nilai paling tinggi, dengan 95% confidence. Berdasarkan hasil hypothesis testing yang telah dilakukan, bisa disimpulkan bahwa harga barang-barang di daerah Jabodetabek lebih mahal dari pada daerah di luar Jabodetabek. Akhirnya, kita juga bisa menyimpulkan bahwa barang-barang yang lebih murah cenderung dibeli orang, dari pada barang-barang yang lebih mahal, dengan perbedaan rata-rata items_sold yaitu 107.
